### ISSUE TITLE
Firefox 56 compatibility

### ISSUE THREAD

**Issue by SammysHP** [vimperator/vimperator-labs#796]
There are some deprecation warnings (not necessary to fix because FF56 will be the last version that is compatible with Vimperator) as well as some major bugs:

**Bugs:**

- [ ] NS_ERROR_ILLEGAL_VALUE: Component returned failure code: 0x80070057 (NS_ERROR_ILLEGAL_VALUE) [nsIProtocolHandler.newChannel]
  This causes #777 and #786.
- [ ] TypeError: services.get(...) is null
- [x] `.textbox-search-icons` renamed to `.textbox-search-sign`

**Warnings:**

- [ ] DEPRECATION WARNING: Search service falling back to synchronous initialization. This is generally the consequence of an add-on using a deprecated search service API.
You may find more details about this deprecation at: https://developer.mozilla.org/en-US/docs/XPCOM_Interface_Reference/nsIBrowserSearchService#async_warning
- [ ] JavaScript 1.6's for-each-in loops are deprecated; consider using ES6 for-of instead
- [ ] String.localeCompare is deprecated; use String.prototype.localeCompare instead
- [ ] String.replace is deprecated; use String.prototype.replace instead
- [ ] String.toLowerCase is deprecated; use String.prototype.toLowerCase instead

**Comment by SammysHP** [comment]
I tried to fix the protocol handler bug for several hours without success. The problem is in `common/components/protocol.js`:

```js
let channel = Components.classesByID["{61ba33c0-3031-11d3-8cd0-0060b0fc14a3}"]
                        .getService(Ci.nsIProtocolHandler)
                        .newChannel(ioService.newURI("chrome://liberator/content/data", null, null))
                        .QueryInterface(Ci.nsIRequest);
```

and fails with `NS_ERROR_ILLEGAL_VALUE: Component returned failure code: 0x80070057 (NS_ERROR_ILLEGAL_VALUE) [nsIProtocolHandler.newChannel]`.

We need help!

**Comment by SammysHP** [comment]
Actually I'm rather happy that this broke! IMHO the chrome-data protocol handler is a big security issue because it might allow websites to gain chrome privileges. But I might be wrong.

**Comment by kyao-777** [comment]
@SammysHP what's the meaning of the FF56 will be the last version that is compatible with Vimperator. You cannot update the vimperator for firefox in the  future. Sorry, my english is bad.

**Comment by SammysHP** [comment]
@kyao-777
Please read #705.

**Comment by kyao-777** [comment]
@SammysHP  thanks

**Comment by JinhaoPro** [comment]
My English is not quite good, but is there anyone who can tell me how to solve #786 problem?

**Comment by SammysHP** [comment]
You have to build an XPI from the ff56-fixes branch yourself and use it with a custom built version of Firefox that does not enforce signature verification. Official update might take a few weeks.

**Comment by maoanran** [comment]
@SammysHP , from Firefox 53, new add-ons can only be WebExtensions, so we can't sign it  ourselves anymore. Sad.

**Comment by P4Cu** [comment]
I'm also highly interested in making that working again so I'm willing to spend some of my time on that but if we're going to split the work I would need an architectural description on what should I change to avoid doing the same stuff by several ppl.

**Comment by SammysHP** [comment]
@maoanran
That's why I recommended to install a version of Firefox that does not require the signature verification.

@P4Cu
Please read #705.

**Comment by SammysHP** [comment]
Just some answers to recent comments:

- Vimperator is semi-maintained currently. There won't be any new features unless someone contributes to it. I try my best to keep Vimperator running, but…

- Firefox 56 is the last version that is compatible with Vimperator. Old extension won't work with FF57 and everything must be programmed again from scratch.

- It seems that nobody with access to addons.mozilla.org is active at the moment. And an updated version may require several weeks for review and might be rejected after all.

- Official versions of Firefox require addons to be signed. Otherwise we would provide a version for direct download.

- **All Bugs under Firefox 58 are fixed**, sadly it is not easy to distribute the fixes, see above.

**Comment by maoanran** [comment]
A signed version for download would be nice. :)

**Comment by SammysHP** [comment]
@maoanran
Only Mozilla can sign an extension and that might not happen anymore.

**Comment by Luftzig** [comment]
@SammysHP I'm trying to understand what can I do to help. If there are tasks that can be tackled by people not familiar with the code base I can suggest this project for the next Goodness Squad meeting in Israel and get several people to work on them for ~3 hours.

**Comment by SammysHP** [comment]
SSL popup is broken again…

**Comment by agush22** [comment]
Any updated xpi?

**Comment by SammysHP** [comment]
> SSL popup is broken again…

Currently it works again… Probably some concurrency issues between Firefox and Vimperator.

**Comment by JulioJu** [comment]
@SammysHP thanks a lot to your work ! It will be very useful for Waterfox 56 (coming soon). 
Maybe, you could publish a new release on https://github.com/vimperator/vimperator-labs/releases and add a link to this Waterfox Release on the top of the README ?

**Comment by JulioJu** [comment]
+1

**Comment by tecfu** [comment]
WORKAROUND:

Please note that Vimperator works fine with Firefox 55. If you are a Ubuntu user like me, you can find Firefox 55 .deb packages [here](https://sourceforge.net/projects/ubuntuzilla/files/mozilla/apt/pool/main/f/firefox-mozilla-build/) at the Ubuntuzilla repository on Sourceforge: 
https://sourceforge.net/projects/ubuntuzilla/files/mozilla/apt/pool/main/f/firefox-mozilla-build/.

**Comment by AnomalRoil** [comment]
No, this is not a solution. Firefox 55 is not a long term support release and should never be used anymore.

On the other hand, [Firefox 52](https://www.mozilla.org/en-US/firefox/organizations/)  is an ESR, which will be supported until August 28. You should use the ESR if you are willing to downgrade to still use Vimperator.
